### PR TITLE
VP-6067: Catalog authentication/authorization scenarios

### PR DESCRIPTION
**Pull request by AliveMen**
## Related task: 
VP-6067 Catalog authentication/authorization scenarios

## Changes:
Make userId nullable for catalog queries

### CODE REVIEW

File: src/VirtoCommerce.ExperienceApiModule.Core/Extensions/ResolveFieldContextExtensions.cs [hunk 1]
```diff
@@ -34,7 +35,14 @@ public static T GetValue<T>(this IResolveFieldContext resolveContext, string key
         public static T GetValue<T>(this IResolveFieldContext resolveContext, string key)
         {
             return resolveContext.GetValue(key, default(T));
-        }     
+        }
+
+        public static string GetCurrentUserId(this IResolveFieldContext resolveContext)
+        {
+            return ((GraphQLUserContext)resolveContext.UserContext).User?.FindFirstValue(ClaimTypes.NameIdentifier)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.Core/Extensions/ResolveFieldContextExtensions.cs:42]: Can you move `((GraphQLUserContext)resolveContext.UserContext).User` to variable?

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs [hunk 2]
```diff
@@ -8,14 +8,15 @@ namespace VirtoCommerce.XDigitalCatalog.Extensions
 {
     public static class ResolveFieldContextExtensions
     {
-        public static T GetCatalogQuery<T>(this IResolveFieldContext context) where T: ICatalogQuery
+        public static T GetCatalogQuery<T>(this IResolveFieldContext context) where T : ICatalogQuery
         {
             var result = AbstractTypeFactory<T>.TryCreateInstance();
             result.StoreId = context.GetArgument<string>("storeId") ?? context.GetValue<string>("storeId");
-            result.UserId = context.GetArgument<string>("userId") ?? context.GetValue<string>("userId");
+            result.UserId = context.GetArgument<string>("userId") ?? context.GetValue<string>("userId") ?? context.GetCurrentUserId();
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs:15]: Could we also move to some extension method `GetArgumentOrValue` - as it seems it uses in many places 
```
context.GetArgument<string>("storeId") ?? context.GetValue<string>("storeId")
```

File: src/VirtoCommerce.ExperienceApiModule.Core/Extensions/ResolveFieldContextExtensions.cs (hunk 1 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.Core/Extensions/ResolveFieldContextExtensions.cs]: Yes, I can!

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs]: added

**Review by yecli — APPROVED** [review]

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs [hunk 3]
```diff
@@ -8,14 +7,15 @@ namespace VirtoCommerce.XDigitalCatalog.Extensions
...
+        public static T GetCatalogQuery<T>(this IResolveFieldContext context) where T : ICatalogQuery
         {
             var result = AbstractTypeFactory<T>.TryCreateInstance();
-            result.StoreId = context.GetArgument<string>("storeId") ?? context.GetValue<string>("storeId");
-            result.UserId = context.GetArgument<string>("userId") ?? context.GetValue<string>("userId");
-            result.CurrencyCode = context.GetArgument<string>("currencyCode") ?? context.GetValue<string>("currencyCode");
-            result.CultureName = context.GetArgument<string>("cultureName") ?? context.GetValue<string>("cultureName");           
-            return result;           
+            result.StoreId = context.GetArgumentOrValue<string>("storeId");
+            result.UserId = context.GetArgumentOrValue<string>("userId") ?? context.GetCurrentUserId();
```
Comment (tatarincev) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ResolveFieldContextExtensions.cs:14]: I think we should get the user id from the session if it doesn't  pass as an argument
https://github.com/VirtoCommerce/vc-module-experience-api/blob/dev/src/XProfile/VirtoCommerce.XProfile/Schemas/ProfileSchema.cs#L44

**Review by yecli — APPROVED** [review]

**Pull request merged**